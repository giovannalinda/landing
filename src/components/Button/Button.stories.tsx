import { Meta, Story } from '@storybook/react'
import { FiHeart } from 'react-icons/fi'
import { Button, ButtonProps } from './Button'

export default {
  title: 'Components/Button',
  component: Button,
  args: {
    as: 'button',
    size: 'md',
    rounded: false,
    variant: 'purple',
  },
} as Meta

export const Default: Story<ButtonProps> = (args) => (
  <Button {...args}>Example</Button>
)

export const Icon: Story<ButtonProps> = (args) => (
  <Button {...args}>
    <FiHeart size={20} />
  </Button>
)
