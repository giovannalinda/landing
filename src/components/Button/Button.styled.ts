import styled, { css } from 'styled-components'
import { theme } from '~/styles'

import { ButtonProps } from './Button'

const sizes = (rounded: boolean) => ({
  sm: css`
    padding: ${theme.space[3]};
    height: 4.8rem;

    ${rounded &&
    css`
      height: 4rem;
      width: 4rem;
    `}
  `,
  md: css`
    padding: ${theme.space[4]} ${theme.space[6]};

    ${rounded &&
    css`
      height: 6rem;
      width: 6rem;
    `}
  `,
})

const variants = {
  purple: css`
    background: ${theme.colors.secondaryPurple};
    color: ${theme.colors.white};

    &:focus {
      outline-offset: 0.3rem;
      outline: 0.2rem solid ${theme.colors.secondaryPurple};
    }
  `,
  white: css`
    background: ${theme.colors.white};
    color: ${theme.colors.primaryDark};

    &:focus {
      outline-offset: 0.3rem;
      outline: 0.2rem solid ${theme.colors.white};
    }
  `,
  outlined: css`
    background: transparent;
    border: 0.1rem solid ${theme.colors.gray100};
    color: ${theme.colors.primaryDark};

    &:hover,
    &:focus {
      color: ${theme.colors.white};
      background: ${theme.colors.secondaryPurple};
    }
  `,
}

export const Container = styled.button<ButtonProps>`
  border-radius: ${theme.borderRadius.button};
  font-weight: ${theme.font.weigths.semiBold};
  letter-spacing: ${theme.font.letterSpacing.sm};
  transition: all ${theme.transitions.fast};
  text-transform: uppercase;

  &:hover {
    transform: translateY(-0.5rem);
    opacity: 0.8;
  }

  ${({ variant }) => variant && variants[variant]};
  ${({ size, rounded = false }) => size && sizes(rounded)[size]};
  ${({ rounded }) =>
    rounded &&
    css`
      padding: 0;
      display: grid;
      place-items: center;
      border-radius: ${theme.borderRadius.rounded};
    `}

  @media screen and (max-width: ${theme.breakpoints.mobile}) {
    font-size: ${theme.font.sizes.small};
    white-space: nowrap;
  }
`
