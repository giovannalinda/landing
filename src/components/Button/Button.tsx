import {
  forwardRef,
  ForwardRefExoticComponent,
  ForwardRefRenderFunction,
  ButtonHTMLAttributes,
  ReactNode,
  AnchorHTMLAttributes,
} from 'react'

import { ButtonAnimation } from './ButtonAnimation'

import * as S from './Button.styled'

type NativeProps =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  as?: 'a' | 'button'
  children: ReactNode
  size?: 'sm' | 'md'
  rounded?: boolean
  variant?: 'purple' | 'white' | 'outlined' | 'neutral'
} & NativeProps

type ButtonCompoundComponet = {
  Animation: typeof ButtonAnimation
} & ForwardRefExoticComponent<ButtonProps>

const ForwardButton: ForwardRefRenderFunction<any, ButtonProps> = (
  { as, children, size = 'md', rounded = false, variant = 'purple', ...rest },
  ref,
) => {
  return (
    <S.Container
      ref={ref}
      as={as}
      size={size}
      rounded={rounded}
      variant={variant}
      {...rest}
    >
      {children}
    </S.Container>
  )
}

export const Button = forwardRef(ForwardButton) as ButtonCompoundComponet

Button.Animation = ButtonAnimation

export * from './useButtonAnimation'
