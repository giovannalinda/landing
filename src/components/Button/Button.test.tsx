import { render, screen } from '@testing-library/react'
import { Button } from './Button'

describe('Button', () => {
  it('should renders correctly', async () => {
    render(<Button>Example</Button>)

    expect(
      await screen.findByRole('button', {
        name: /Example/i,
      }),
    ).toBeInTheDocument()
  })

  it('should match snapshot', () => {
    const { container } = render(<Button>Example</Button>)

    expect(container.firstChild).toMatchSnapshot()
  })
})
