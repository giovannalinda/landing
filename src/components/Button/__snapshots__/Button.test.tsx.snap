// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<button
  class="sc-bczRLJ bRsxPm"
>
  Example
</button>
`;
